meta/autoid/autoid_test.go

```diff
@@ -558,9 +558,214 @@ func BenchmarkAllocator_Alloc(b *testing.B) {
+	// the rest in cache in not enough for next value.
+	c.Assert(ok, Equals, false)
+
+	base, end, round, err = alloc.AllocSeqCache(1)
+	c.Assert(err, IsNil)
+	c.Assert(base, Equals, int64(-11))
+	c.Assert(end, Equals, int64(-6))
+	// the round already in cycle.
+	c.Assert(round, Equals, int64(1))
+
+	// test the sequence batch size.
+	size, err = autoid.CalcSequenceBatchSize(sequenceBase, seq.CacheValue, seq.Increment, offset, seq.MinValue, seq.MaxValue)
+	c.Assert(err, IsNil)
+	c.Assert(end-base, Equals, size)
+
+	offset = seq.MinValue
+	nextVal, ok = autoid.SeekToFirstSequenceValue(base, seq.Increment, offset, base, end)
+	c.Assert(ok, Equals, true)
+	c.Assert(nextVal, Equals, int64(-10))
+	base = nextVal
+
+	nextVal, ok = autoid.SeekToFirstSequenceValue(base, seq.Increment, offset, base, end)
+	c.Assert(ok, Equals, true)
+	c.Assert(nextVal, Equals, int64(-8))
+	base = nextVal
+
+	nextVal, ok = autoid.SeekToFirstSequenceValue(base, seq.Increment, offset, base, end)
+	c.Assert(ok, Equals, true)
+	c.Assert(nextVal, Equals, int64(-6))
+	base = nextVal
+
+	_, ok = autoid.SeekToFirstSequenceValue(base, seq.Increment, offset, base, end)
+	// the cache is already empty.
+	c.Assert(ok, Equals, false)
+}
+
+func (*testSuite) TestConcurrentAllocSequence(c *C) {
+	store, err := mockstore.NewMockTikvStore()
+	c.Assert(err, IsNil)
+	defer store.Close()
+
+	var seq *model.SequenceInfo
+	var sequenceBase int64
+	err = kv.RunInNewTxn(store, false, func(txn kv.Transaction) error {
+		m := meta.NewMeta(txn)
+		err1 := m.CreateDatabase(&model.DBInfo{ID: 2, Name: model.NewCIStr("a")})
+		c.Assert(err1, IsNil)
+		seq = &model.SequenceInfo{
+			Start:      100,
+			Cycle:      false,
+			Cache:      true,
+			MinValue:   -100,
+			MaxValue:   100,
+			Increment:  -2,
+			CacheValue: 3,
+		}
+		seqTable := &model.TableInfo{
+			ID:       2,
+			Name:     model.NewCIStr("seq"),
+			Sequence: seq,
+		}
+		if seq.Increment >= 0 {
+			sequenceBase = seq.Start - 1
+		} else {
+			sequenceBase = seq.Start + 1
+		}
+		err1 = m.CreateSequenceAndSetSeqValue(2, seqTable, sequenceBase)
+		c.Assert(err1, IsNil)
+		return nil
+	})
+	c.Assert(err, IsNil)
+
+	var mu sync.Mutex
+	wg := sync.WaitGroup{}
+	m := map[int64]struct{}{}
+	count := 10
+	errCh := make(chan error, count)
+
+	allocSequence := func() {
+		alloc := autoid.NewAllocator(store, 2, false, autoid.SequenceType, seq)
+		for j := 0; j < 3; j++ {
+			base, end, _, err1 := alloc.AllocSeqCache(2)
+			if err1 != nil {
+				errCh <- err1
+				break
+			}
+
+			errFlag := false
+			mu.Lock()
+			// sequence is negative-growth here.
+			for i := base - 1; i >= end; i-- {
+				if _, ok := m[i]; ok {
+					errCh <- fmt.Errorf("duplicate id:%v", i)
+					errFlag = true
+					mu.Unlock()
+					break
+				}
+				m[i] = struct{}{}
+			}
+			if errFlag {
+				break
+			}
+			mu.Unlock()
+		}
+	}
+	for i := 0; i < count; i++ {
+		wg.Add(1)
+		go func(num int) {
+			defer wg.Done()
```

Comment:
We'd better move `wg.Done()` to the end of this function, and avoid using `defer`.